File: sdk/python/tempmail_sdk/providers/mail_gw.py
"""
mail.gw 渠道实现
API 与 mail.tm 完全一致，仅 baseURL 不同
"""

import random
import string
from ..types import EmailInfo
from ..normalize import normalize_email
from .. import http as tm_http

CHANNEL = "mail-gw"
BASE_URL = "https://api.mail.gw"

DEFAULT_HEADERS = {
    "Content-Type": "application/json",
    "Accept": "application/json",
}


def _random_string(length):
    return "".join(random.choices(string.ascii_lowercase + string.digits, k=length))


def _get_domains():
    resp = tm_http.get(f"{BASE_URL}/domains", headers=DEFAULT_HEADERS)
    resp.raise_for_status()
    data = resp.json()
    members = data if isinstance(data, list) else data.get("hydra:member", [])
    return [d["domain"] for d in members if d.get("isActive")]


def _create_account(address, password):
    resp = tm_http.post(
        f"{BASE_URL}/accounts",
        headers={**DEFAULT_HEADERS, "Content-Type": "application/ld+json"},
        json={"address": address, "password": password},
    )
    resp.raise_for_status()
    return resp.json()


def _get_token(address, password):
    resp = tm_http.post(
        f"{BASE_URL}/token",
        headers=DEFAULT_HEADERS,
        json={"address": address, "password": password},
    )
    resp.raise_for_status()
    return resp.json()["token"]


def generate_email():
    """创建临时邮箱"""
    domains = _get_domains()
    if not domains:
        raise Exception("No available domains")

    domain = random.choice(domains)
    username = _random_string(12)
    address = f"{username}@{domain}"
    password = _random_string(16)

    account = _create_account(address, password)
    token = _get_token(address, password)

    return EmailInfo(
        channel=CHANNEL,
        email=address,
        _token=token,
        created_at=account.get("createdAt"),
    )


def _flatten_message(msg, recipient_email):
    return {
        "id": msg.get("id", ""),
        "from": (msg.get("from") or {}).get("address", ""),
        "to": (msg.get("to") or [{}])[0].get("address", "") or recipient_email,
        "subject": msg.get("subject", ""),
        "text": msg.get("text", ""),
        "html": "".join(msg["html"]) if isinstance(msg.get("html"), list) else msg.get("html", ""),
        "createdAt": msg.get("createdAt", ""),
        "seen": msg.get("seen", False),
        "attachments": [
            {
                "filename": a.get("filename", ""),
                "size": a.get("size"),
                "contentType": a.get("contentType"),
                "downloadUrl": f"{BASE_URL}{a['downloadUrl']}" if a.get("downloadUrl") else None,
            }
            for a in msg.get("attachments", [])
        ],
    }


def get_emails(token, email="", **kwargs):
    """获取邮件列表"""
    resp = tm_http.get(
        f"{BASE_URL}/messages",
        headers={**DEFAULT_HEADERS, "Authorization": f"Bearer {token}"},
    )
    resp.raise_for_status()
    data = resp.json()
    messages = data if isinstance(data, list) else data.get("hydra:member", [])

    if not messages:
        return []

    emails = []
    for msg in messages:
        try:
            detail_resp = tm_http.get(
                f"{BASE_URL}/messages/{msg['id']}",
                headers={**DEFAULT_HEADERS, "Authorization": f"Bearer {token}"},
            )
            detail_resp.raise_for_status()
            detail = detail_resp.json()
            flat = _flatten_message(detail, email)
        except Exception:
            flat = _flatten_message(msg, email)
        emails.append(normalize_email(flat, email))

    return emails

File: sdk/npm/src/providers/mail-gw.ts
import { InternalEmailInfo, Email, Channel } from '../types';
import { normalizeEmail } from '../normalize';
import { fetchWithTimeout } from '../retry';
import { randomInt } from 'crypto';

const CHANNEL: Channel = 'mail-gw';
const BASE_URL = 'https://api.mail.gw';
const DEFAULT_HEADERS: Record<string, string> = {
  'Accept': 'application/json',
  'Accept-Language': 'zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6',
  'Cache-Control': 'no-cache',
  'Pragma': 'no-cache',
  'Origin': 'https://mail.gw',
  'Referer': 'https://mail.gw/',
  'User-Agent':
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0',
};

function jsonHeaders(extra?: Record<string, string>): Record<string, string> {
  return { ...DEFAULT_HEADERS, 'Content-Type': 'application/json', ...extra };
}

function bearerHeaders(token: string, extra?: Record<string, string>): Record<string, string> {
  return { ...jsonHeaders(extra), Authorization: `Bearer ${token}` };
}

/**
 * 解析 Hydra 格式或普通数组响应
 */
function parseHydraMember(data: unknown): any[] {
  if (Array.isArray(data)) return data;
  if (data && typeof data === 'object') {
    const o = data as Record<string, unknown>;
    if (Array.isArray(o['hydra:member'])) return o['hydra:member'];
    if (Array.isArray(o['data'])) return o['data'];
  }
  return [];
}

/**
 * 生成随机字符串
 */
function randomString(length: number): string {
  const chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
  let result = '';
  for (let i = 0; i < length; i++) {
    result += chars[randomInt(chars.length)];
  }
  return result;
}

/**
 * 获取可用域名列表
 * API: GET /domains
 */
async function getDomains(): Promise<string[]> {
  const response = await fetchWithTimeout(`${BASE_URL}/domains`, {
    method: 'GET',
    headers: DEFAULT_HEADERS,
  });

  if (!response.ok) {
    throw new Error(`获取域名列表失败: ${response.status}`);
  }

  const data = await response.json();
  const members = parseHydraMember(data);
  return members
    .filter((d: any) => d && d.isActive !== false)
    .map((d: any) => d.domain)
    .filter(Boolean);
}

/**
 * 创建账号
 * API: POST /accounts
 */
async function createAccount(address: string, password: string): Promise<any> {
  const response = await fetchWithTimeout(`${BASE_URL}/accounts`, {
    method: 'POST',
    headers: jsonHeaders(),
    body: JSON.stringify({ address, password }),
  });

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`创建账号失败: ${response.status} ${text}`);
  }

  return response.json();
}

/**
 * 获取 Bearer Token
 * API: POST /token
 */
async function getToken(address: string, password: string): Promise<string> {
  const response = await fetchWithTimeout(`${BASE_URL}/token`, {
    method: 'POST',
    headers: jsonHeaders(),
    body: JSON.stringify({ address, password }),
  });

  if (!response.ok) {
    throw new Error(`获取令牌失败: ${response.status}`);
  }

  const data = await response.json();
  return data.token;
}

/**
 * 生成临时邮箱
 * 流程：获取域名 → 随机选择域名 → 创建账号 → 登录获取 JWT → 返回邮箱信息
 */
export async function generateEmail(): Promise<InternalEmailInfo> {
  // 1. 获取可用域名
  const domains = await getDomains();
  if (domains.length === 0) {
    throw new Error('没有可用的域名');
  }

  // 2. 随机选择域名，生成地址和密码
  const domain = domains[randomInt(domains.length)];
  const username = randomString(12);
  const address = `${username}@${domain}`;
  const password = randomString(16);

  // 3. 创建账号
  const account = await createAccount(address, password);

  // 4. 获取 Bearer Token
  const token = await getToken(address, password);

  return {
    channel: CHANNEL,
    email: address,
    token: token,
    createdAt: account.createdAt,
  };
}

/**
 * 将消息扁平化为统一格式
 */
function flattenMessage(msg: any, recipientEmail: string): any {
  return {
    id: msg.id || '',
    from: msg.from?.address || '',
    to: msg.to?.[0]?.address || recipientEmail,
    subject: msg.subject || '',
    text: msg.text || '',
    html: Array.isArray(msg.html) ? msg.html.join('') : (msg.html || ''),
    createdAt: msg.createdAt || '',
    seen: msg.seen ?? false,
    attachments: (msg.attachments || []).map((a: any) => ({
      filename: a.filename || '',
      size: a.size || undefined,
      contentType: a.contentType || undefined,
      downloadUrl: a.downloadUrl ? `${BASE_URL}${a.downloadUrl}` : undefined,
    })),
  };
}

/**
 * 获取邮件列表
 * 使用 JWT 令牌获取消息列表，并逐条获取详情以获得完整的邮件正文
 */
export async function getEmails(token: string, email: string): Promise<Email[]> {
  // 获取消息列表
  const listResponse = await fetchWithTimeout(`${BASE_URL}/messages`, {
    method: 'GET',
    headers: bearerHeaders(token),
  });

  if (!listResponse.ok) {
    throw new Error(`获取邮件列表失败: ${listResponse.status}`);
  }

  const listData = await listResponse.json();
  const messages = parseHydraMember(listData);

  if (messages.length === 0) {
    return [];
  }

  // 逐条获取邮件详情（包含完整正文）
  const detailPromises = messages.map(async (msg: any) => {
    try {
      const detailResponse = await fetchWithTimeout(`${BASE_URL}/messages/${msg.id}`, {
        method: 'GET',
        headers: bearerHeaders(token),
      });

      if (!detailResponse.ok) {
        return flattenMessage(msg, email);
      }

      const detail = await detailResponse.json();
      return flattenMessage(detail, email);
    } catch {
      return flattenMessage(msg, email);
    }
  });

  const flatMessages = await Promise.all(detailPromises);
  return flatMessages.map((raw: any) => normalizeEmail(raw, email));
}
